### ISSUE TITLE
[BUG] IIS check on Windows generates event logs

### ISSUE THREAD

**Issue by SanderKnape** [DataDog/datadog-agent#13605]
**Agent Environment**
7.39.0
Agent logs show nothing of relevance

**Describe what happened:**
Installed Datadog Agent 7.39.0 using the Puppet module
Enabled IIS check
Windows Event Logs appear:

```
Log Name:      Application
Source:        Microsoft-Windows-Perflib
Date:          9/21/2022 9:43:33 AM
Event ID:      1008
Task Category: None
Level:         Error
Keywords:      Classic
User:          N/A
Computer:      skttew1dev-web2
Description:
The Open Procedure for service "WmiApRpl" in DLL "C:\Windows\system32\wbem\wmiaprpl.dll" failed. Performance data for this service will not be available. The first four bytes (DWORD) of the Data section contains the error code.
Event Xml:
<Event xmlns="http://schemas.microsoft.com/win/2004/08/events/event">
  <System>
    <Provider Name="Microsoft-Windows-Perflib" Guid="{13B197BD-7CEE-4B4E-8DD0-59314CE374CE}" EventSourceName="Perflib" />
    <EventID Qualifiers="49152">1008</EventID>
    <Version>0</Version>
    <Level>2</Level>
    <Task>0</Task>
    <Opcode>0</Opcode>
    <Keywords>0x80000000000000</Keywords>
    <TimeCreated SystemTime="2022-09-21T09:43:33.687236700Z" />
    <EventRecordID>210522</EventRecordID>
    <Correlation />
    <Execution ProcessID="0" ThreadID="0" />
    <Channel>Application</Channel>
    <Computer>skttew1dev-web2</Computer>
    <Security />
  </System>
  <UserData>
    <EventXML xmlns="Perflib">
      <param1>WmiApRpl</param1>
      <param2>C:\Windows\system32\wbem\wmiaprpl.dll</param2>
      <binaryDataSize>8</binaryDataSize>
      <binaryData>1500000000000000</binaryData>
    </EventXML>
  </UserData>
</Event>
```

```
Log Name:      Application
Source:        Microsoft-Windows-Perflib
Date:          9/21/2022 9:43:33 AM
Event ID:      1008
Task Category: None
Level:         Error
Keywords:      Classic
User:          N/A
Computer:      skttew1dev-web2
Description:
The Open Procedure for service "BITS" in DLL "C:\Windows\System32\bitsperf.dll" failed. Performance data for this service will not be available. The first four bytes (DWORD) of the Data section contains the error code.
Event Xml:
<Event xmlns="http://schemas.microsoft.com/win/2004/08/events/event">
  <System>
    <Provider Name="Microsoft-Windows-Perflib" Guid="{13B197BD-7CEE-4B4E-8DD0-59314CE374CE}" EventSourceName="Perflib" />
    <EventID Qualifiers="49152">1008</EventID>
    <Version>0</Version>
    <Level>2</Level>
    <Task>0</Task>
    <Opcode>0</Opcode>
    <Keywords>0x80000000000000</Keywords>
    <TimeCreated SystemTime="2022-09-21T09:43:33.608589100Z" />
    <EventRecordID>210520</EventRecordID>
    <Correlation />
    <Execution ProcessID="0" ThreadID="0" />
    <Channel>Application</Channel>
    <Computer>skttew1dev-web2</Computer>
    <Security />
  </System>
  <UserData>
    <EventXML xmlns="Perflib">
      <param1>BITS</param1>
      <param2>C:\Windows\System32\bitsperf.dll</param2>
      <binaryDataSize>8</binaryDataSize>
      <binaryData>0500000000000000</binaryData>
    </EventXML>
  </UserData>
</Event>
```

```
Log Name:      Application
Source:        Microsoft-Windows-PerfNet
Date:          9/21/2022 9:43:33 AM
Event ID:      2004
Task Category: None
Level:         Error
Keywords:      Classic
User:          N/A
Computer:      skttew1dev-web2
Description:
Unable to open the Server service performance object. The first four bytes (DWORD) of the Data section contains the status code.
Event Xml:
<Event xmlns="http://schemas.microsoft.com/win/2004/08/events/event">
  <System>
    <Provider Name="Microsoft-Windows-PerfNet" Guid="{CAB2B8A5-49B9-4EEC-B1B0-FAC21DA05A3B}" EventSourceName="PerfNet" />
    <EventID Qualifiers="49152">2004</EventID>
    <Version>0</Version>
    <Level>2</Level>
    <Task>0</Task>
    <Opcode>0</Opcode>
    <Keywords>0x80000000000000</Keywords>
    <TimeCreated SystemTime="2022-09-21T09:43:33.655424500Z" />
    <EventRecordID>210521</EventRecordID>
    <Correlation />
    <Execution ProcessID="0" ThreadID="0" />
    <Channel>Application</Channel>
    <Computer>skttew1dev-web2</Computer>
    <Security />
  </System>
  <EventData>
    <Binary>220000C0</Binary>
  </EventData>
</Event>
```

**Describe what you expected:**
No error logs in Windows Event Log

**Steps to reproduce the issue:**
Install IIS on web server
Install agent 7.39.0
Enable IIS Check

**Additional environment details (Operating System, Cloud provider, etc):**
Windows Server 2016 Datacenter version 1607, OS Build 14393.5291 running in Azure

This issue looks very similar to https://github.com/DataDog/datadog-agent/issues/11145, except the errors in the Datadog agent logs are no longer present.

Downgrading to version 7.33.1 ([as mentioned before](https://github.com/DataDog/datadog-agent/issues/12899#issuecomment-1198795708)) removes the errors.

**Comment by iglendd** [comment]
@SanderKnape Thank you for reporting this issue. Technically it is caused by and should be filed with https://github.com/DataDog/integrations-core. We have already identified its root cause and are working to address it.

Event Log warnings and errors are reported because we have started using the Performance Counter PdhEnumObjects() API in the Agent 7.34. When this API is invoked it will try to interrogate all performance counter objects code and some of them fail and report it as an error or warning to the Event Log if the current process is running as a non-administrator. This is not documented Windows behavior because the agent service user is added to the required "Performance Monitor Users" group. Besides polluting Event Log these warnings and errors do not cause any harm or indicate any misbehavior by the Datadog agent or other software.

Until we have a fix to this nuisance you can consider one of the following workarounds:

- Run agent service as Local System service.
- Add agent users to the local Administrators group.
- In pendant Agent 7.40 release one can configure refreshment frequency (windows_counter_refresh_interval: N - in seconds) or completely suppress it when  windows_counter_refresh_interval: 0 in datadog.yaml
- Explicitly disable offending Performance Counter(s), e.g. PerfNet and WMIAprPl. These Performance Counters generally are not often used and can be suppressed. Microsoft provides documentation for that (https://docs.microsoft.com/en-us/troubleshoot/windows-server/performance/manually-rebuild-performance-counters#ensure-that-the-counters-arent-disabled-in-the-registry, https://docs.microsoft.com/en-us/previous-versions/windows/it-pro/windows-server-2003/cc737243(v=ws.10), https://techcommunity.microsoft.com/t5/ask-the-performance-team/two-minute-drill-disabled-performance-counters-and-exctrlst-exe/ba-p/374538

**Comment by SanderKnape** [comment]
Thanks for the information @iglendd ! Good to know that it is being looked into to and that apart from the error messages no harm is done on the system. Once a new version with a fix is released, we'll give it a go.

**Comment by iglendd** [comment]
@SanderKnape  the fix is tentatively scheduled to be in the Agent 7.42 release.

As a side note, Microsoft support has confirmed that while membership in the "Performance Monitor Users" group is needed to consume performance counters, technically, it is not enough for all possible counters because some of them do require elevated privileges, e.g. SCCM counters or Computer Browser counters. It certainly causes benign but annoying side effects of generating Application and Security Event log errors, audit failures and warnings, without impacting accessibility to the most of the performance counters which are not requiring elevated privileges.

**Comment by SanderKnape** [comment]
Thanks for the update @iglendd. Looking forward to give this a try.

**Comment by iglendd** [comment]
@SanderKnape 7.42 is release and contain the fix which you can try

**Comment by clarkb7** [comment]
Closing since we haven't heard back but think this issue is fixed in Agent 7.42. If you're still having issues please re-open the issue or [contact support](https://www.datadoghq.com/support/).
